torchlpc/csrc/scan_cpu.cpp

```diff
@@ -0,0 +1,82 @@
+#include <torch/script.h>
+#include <torch/torch.h>
+#include <algorithm>
+#include <vector>
+#include <utility>
+
+template <typename scalar_t>
+void scan_cpu(const at::Tensor &input, const at::Tensor &weights, const at::Tensor &initials, at::Tensor &output)
+{
+    TORCH_CHECK(input.dim() == 2, "Input must be 2D");
+    TORCH_CHECK(initials.dim() == 1, "Initials must be 1D");
+    TORCH_CHECK(weights.sizes() == input.sizes(), "Weights must have the same size as input");
+    TORCH_CHECK(output.sizes() == input.sizes(), "Output must have the same size as input");
+    TORCH_CHECK(initials.size(0) == input.size(0), "The first dimension of initials must be the same as the first dimension of input");
+    TORCH_INTERNAL_ASSERT(input.device().is_cpu(), "Input must be on CPU");
+    TORCH_INTERNAL_ASSERT(initials.device().is_cpu(), "Initials must be on CPU");
+    TORCH_INTERNAL_ASSERT(weights.device().is_cpu(), "Weights must be on CPU");
+    TORCH_INTERNAL_ASSERT(output.device().is_cpu(), "Output must be on CPU");
+    // TORCH_INTERNAL_ASSERT(input.is_contiguous(), "Input must be contiguous");
+    // TORCH_INTERNAL_ASSERT(weights.is_contiguous(), "Weights must be contiguous");
+    TORCH_INTERNAL_ASSERT(output.is_contiguous(), "Output must be contiguous");
+
+    auto input_contiguous = input.contiguous();
+    auto weights_contiguous = weights.contiguous();
+    auto initials_contiguous = initials.contiguous();
```

Comment:
Why not make the output contiguous as well here since the data in it will be overwritten anyways with the output values?

Reply:
The output is created outside the scope of this function and should be the user's responsibility to make sure it's contiguous. Also `output.contiguous()` might not point to the original tensor.